pkg/url/url.go

```diff
@@ -512,16 +542,46 @@ func ListIngressAndRoute(oclient *occlient.Client, client *kclient.Client, envSp
 		clusterURL := getMachineReadableFormat(r)
 		clusterURLMap[clusterURL.Name] = clusterURL
 	}
-	for _, envinfoURL := range localEnvinfoURLs {
-		// only checks for Ingress and Route URLs
-		if envinfoURL.Kind == envinfo.DOCKER {
-			continue
-		}
-		if !routeSupported && envinfoURL.Kind == envinfo.ROUTE {
-			continue
+	for _, endpointMap := range containerEndpointsMap {
+		for _, localEndpoint := range endpointMap {
+			// only exposed endpoint will be shown as a URL in `odo url list`
+			if localEndpoint.Exposure == common.None || localEndpoint.Exposure == common.Internal {
+				continue
+			}
+			var devfileURL envinfo.EnvInfoURL
+			for _, envinfoURL := range localEnvinfoURLs {
```

Comment:
put these URLs in a map and fetch from it

Reply:
Why does it matter? Convert a slice into a map needs to iterating through the entire slice. 
We can break the loop once we find the matching name when looping through the slice inside.

Reply:
This is currently a O(n3) without any breaks. We can break after finding the URL, but if the endpoint, we are looking for, is in the last container of containerEndpointsMap and the last endpoint of the corresponding container, we will unnecessarily loop over localEnvinfoURLs multiple times. Thus looping over localEnvinfoURLs once and storing it once in a map is much more efficient and prevents the unnecessary iterations even in the worst case scenario.

Reply:
updated